Include/cpython/lock.h

```diff
@@ -30,32 +30,34 @@ typedef struct PyMutex {
     uint8_t _bits;  // (private)
 } PyMutex;
 
-// (private) slow path for locking the mutex
-PyAPI_FUNC(void) _PyMutex_LockSlow(PyMutex *m);
+// exported function for locking the mutex
+PyAPI_FUNC(void) PyMutex_Lock(PyMutex *m);
 
-// (private) slow path for unlocking the mutex
-PyAPI_FUNC(void) _PyMutex_UnlockSlow(PyMutex *m);
+// exported function for unlocking the mutex
+PyAPI_FUNC(void) PyMutex_Unlock(PyMutex *m);
 
 // Locks the mutex.
 //
 // If the mutex is currently locked, the calling thread will be parked until
 // the mutex is unlocked. If the current thread holds the GIL, then the GIL
 // will be released while the thread is parked.
 static inline void
-PyMutex_Lock(PyMutex *m)
+_PyMutex_Lock(PyMutex *m)
 {
     uint8_t expected = _Py_UNLOCKED;
     if (!_Py_atomic_compare_exchange_uint8(&m->_bits, &expected, _Py_LOCKED)) {
-        _PyMutex_LockSlow(m);
+        PyMutex_Lock(m);
```

Comment:
Why not keeping _PyMutex_LockSlow()? Is it a recursive call? It's unclear to me.

I would expect PyMutex_Lock() to the exactly the same than the static inline function.

Reply:
It doesn't matter much, but it's convenient to make the "slow path" and the non-inline function the same function.

The inline function (`_PyMutex_Lock`) is written to generate few instructions because it's inlined in lots of places, but that consideration doesn't make as much sense for the non-inline function.

Reply:
To be more clear about this, the non-inlined function starts with a more general handling of `unlocked -> locked` transition that also handles the case where there's a waiter (but the lock is unlocked):

https://github.com/python/cpython/blob/35b16795d11cb50768ffad5fe8e61bdebde9b66a/Python/lock.c#L59-L64

The inline function only handles the specific case where the lock is unlocked AND there are no waiters:

https://github.com/python/cpython/blob/35b16795d11cb50768ffad5fe8e61bdebde9b66a/Include/internal/pycore_lock.h#L73-L74

The more general case is ~3 extra instructions. It's usually worth avoiding that in the inline path, but that consideration doesn't really matter in the non-inline case.

Reply:
Ok I see. It's surprising, but it makes sense. Also, it's up to you :-)